src/main/java/org/jboss/threads/EnhancedQueueExecutor.java

```diff
@@ -1515,16 +1522,33 @@ private QNode getOrAddNode(PoolThreadNode nextPoolThreadNode) {
             for (;;) {
                 head = EnhancedQueueExecutor.this.head;
                 headNext = head.getNext();
+                // this happen if another consumer has already consumed head:
+                // retry with a fresh head
+                if (headNext == head) {
+                    if (UPDATE_STATISTICS) spinMisses.increment();
+                    JDKSpecific.onSpinWait();
+                    continue;
+                }
                 if (headNext instanceof TaskNode) {
                     TaskNode taskNode = (TaskNode) headNext;
                     if (compareAndSetHead(head, taskNode)) {
+                        // save from GC Nepotism: generational GCs don't like
+                        // cross-generational references, so better to "clean-up" head::next
+                        // to save dragging head::next into the old generation.
+                        // Clean-up cannot just null out next
+                        head.setNextOrdered(head);
                         if (! NO_QUEUE_LIMIT) decreaseQueueSize();
                         return taskNode;
                     }
                 } else if (headNext instanceof PoolThreadNode || headNext == null) {
                     nextPoolThreadNode.setNextRelaxed(headNext);
                     if (head.compareAndSetNext(headNext, nextPoolThreadNode)) {
                         return nextPoolThreadNode;
+                    } else if (headNext != null) {
+                        // GC Nepotism:
+                        // save dragging headNext into old generation
+                        // (although being a PoolThreadNode it won't make a big difference)
+                        nextPoolThreadNode.setNextRelaxed(null);
```

Comment:
Am I correct in understanding that this is presumed safe because `nextPoolThreadNode` has not yet been published?

Reply:
Excactly and because `nextPoolThreadNode` (if not published) is still owned by a single thread (the consumer thread itself), hence `setNextRelaxed` is fine,because no other threads would never be able to read `nextPoolThreadNode::next` concurrently.